### DISCUSSION TITLE
Support new Inline Auto-Fill Option in Browser Extension

### DISCUSSION THREAD

**Discussion by 1h8fulkat** [dani-garcia/vaultwarden#4254] [Ideas]
https://bitwarden.com/help/auto-fill-browser/#enable-the-inline-auto-fill-menu

Can we enable Vaultwarden to use of inline auto-fill on the Bitwarden browser extension?

**Comment by BlackDex** [comment]
If you use the `:testing` tagged images you already can.
https://github.com/dani-garcia/vaultwarden/blob/890e668071cffe2833834348e19bbef3c061d014/.env.template#L395-L404

**Reply by tessus** [reply]
btw, which of the `autofill-*` flags are set by default in Bitwarden (sorry, I couldn't find the defaults in the bw code, which is not so easy to read as the vw code)? I remember that there was an issue with autofill at one point, but I don't recall the specifics.

**Reply by 1h8fulkat** [reply]
@BlackDex  I've tried both the testing tag as well as setting the "EXPERIMENTAL_CLIENT_FEATURE_FLAGS" variable to "autofill-v2" in both my docker compose file, config.json and .env file and I can't get that feature to appear in the auto-fill menu of the Chrome extension.

**Reply by miandru** [reply]
Try deleting the browser extension and adding it again.
The environment value should be
`EXPERIMENTAL_CLIENT_FEATURE_FLAGS=autofill-overlay,autofill-v2`

**Reply by stefan0xC** [reply]
You need both `autofill-v2` and `autofill-overlay` to enable this new option. 

@tessus you can look up the `/api/config` endpoint to see what the bitwarden cloud uses (which differs a bit from the defaults for self-hosted IIRC). Be aware that we haven't implemented the needed server-side changes for all enabled features in Vaultwarden yet.

**Reply by tessus** [reply]
Thanks @stefan0xC, I know that the endpoints for passphrase login still have to be implemented. However, the feature flag for that is not in the `KNOWN_FLAGS` array, thus it can't even be set in vw.

**Reply by stefan0xC** [reply]
@tessus I've just tested it and you can set any flag you want. I've fixed this in https://github.com/dani-garcia/vaultwarden/pull/4263

**Reply by tessus** [reply]
Ah, I never tried, I only read the code... and apparently not very thoroughly. ;-)

**Comment by 1h8fulkat** [comment]
> You need both `autofill-v2` and `autofill-overlay` to enable this new option.

Thanks!  That seems to have done the trick.

**Comment by dfunkt** [comment]
Looks like the flags for autofill-v2 and autofill-overlay are going to be removed in bitwarden's server:
https://github.com/bitwarden/server/pull/3692

**Reply by stefan0xC** [reply]
Makes sense because they have been removed in the web-vault: https://github.com/bitwarden/clients/commit/0127db44a00eaf27378d899d0c44aa255fd05a75
This just means that users won't have to enable them manually once a new web-vault with this change has been released. (Since this is a client-side only feature we don't have to do anything, except maybe removing the feature flag from `KNOWN_FLAGS` eventually.)

**Comment by lukasj98** [comment]
When is this option coming to main release version

**Reply by BlackDex** [reply]
This was already for a while in the main/testing releases.

If you mean the stable/latest release, then there is no schedule or release cadence, you just have to wait until we deem it ready to be called a stable.

**Reply by lukasj98** [reply]
Can I use inline auto fill already

**Reply by BlackDex** [reply]
See the env template: https://github.com/dani-garcia/vaultwarden/blob/4b9384cb2bdc98d6805e91f73a0ed6ed8ddfbec3/.env.template#L403-L408
And add both autofill items to the list.

**Reply by bibicadotnet** [reply]
use `image:vaultwarden/server:testing`
Add environment:
`- EXPERIMENTAL_CLIENT_FEATURE_FLAGS=autofill-overlay,autofill-v2,browser-fileless-import,fido2-vault-credentials`

The Inline Auto-Fill feature has many errors, it cannot automatically copy the 2FA code, and also causes errors in certain forms.

Turned off as default vaultwarden is used more effectively, when using just use the shortcut Ctrl + Shift + T more effectively

**Reply by tessus** [reply]
> The Inline Auto-Fill feature has many errors, it cannot automatically copy the 2FA code, and also causes errors in certain forms.

I have also noticed the issue with 2FA tokens not being copied. However, these are client features and have to be additionally activated on the client anyway. Or at least for now.

Thus setting it on the server does not hurt, but rather gives the users the option to activate it whenver they want.

I hope that BW is aware of these issues and is working on fixing the clients.

**Reply by bibicadotnet** [reply]
On the official Bitwarden, even if you completely turn off the Inline Auto-Fill feature from the client, it will still cause errors on some pages when using Ctrl + Shift + T

On vaultwarden, it can be turned off completely from the server, so it doesn't cause this problem

**Reply by tessus** [reply]
Hmm, I turned it off in the client (vw server still has `autofill-overlay` and `autofill-v2` set) and all of a sudden copying the 2FA token works again.

**Reply by bibicadotnet** [reply]
As I said above, this feature conflicts with some forms, specifically you can watch through the [video](https://youtu.be/wIi-pLdipu8?si=JV6lCbzDJy5ETRse)
Even though I turned off all features related to Inline Auto-Fill, it still causes errors, even on the official Bitwarden server version.
On Vaultwarden, when this feature is turned off from the server, everything works normally

**Reply by tessus** [reply]
Maybe I expressed myself poorly.

On Vaultwarden, when this feature is turned on on the server, but disabled in the clients, everything works normally.
I didn't test hundreds of forms, but the ones I tested worked.

**Reply by bibicadotnet** [reply]
~~It's quite confusing, because I tried Bitwarden Unified self-hosted v2024.1.2, the Inline Auto-Fill feature is not available on this version, but it also does not cause errors for some of the forms I am using.~~

In general, Bitwarden's Inline Auto-Fill feature is temporarily ineffective. It's best to just use it as before, using `Ctrl + Shift + T` is enough.

Ah, I would like to correct myself a bit, I just reinstalled Bitwarden Unified self-hosted v2024.1.2 and no longer had the error problem related to Inline Auto-Fill.

**Reply by lukasj98** [reply]
> ~~It's quite confusing, because I tried Bitwarden Unified self-hosted v2024.1.2, the Inline Auto-Fill feature is not available on this version, but it also does not cause errors for some of the forms I am using.~~
> 
> 
> 
> In general, Bitwarden's Inline Auto-Fill feature is temporarily ineffective. It's best to just use it as before, using `Ctrl + Shift + T` is enough.
> 
> 
> 
> Ah, I would like to correct myself a bit, I just reinstalled Bitwarden Unified self-hosted v2024.1.2 and no longer had the error problem related to Inline Auto-Fill.

Is inline auto fill available in self host Bitwarden

**Reply by BlackDex** [reply]
The feature flags are not there. So that means that the clients use there default, which is different per version.
It still seems to be a feature flag in v2024.1.2, but in main it has been removed.
v2024.1.2: https://github.com/bitwarden/clients/blob/web-v2024.1.2/libs/common/src/enums/feature-flag.enum.ts
main: https://github.com/bitwarden/clients/blob/main/libs/common/src/enums/feature-flag.enum.ts

So probably the next version, it doesn't matter if it is there or not.

That means, no, in self-hosted it is not available, since the clients are not triggered to enable it.
Checkout your `https://self-hosted-bitwarden.domain.tld/api/config` and for there hosted one see https://vault.bitwarden.com/api/config or https://api.bitwarden.com/config

**Reply by bibicadotnet** [reply]
> Is inline auto fill available in self host Bitwarden

I'm not sure if you're mistakenly referring to the official version that Bitwarden uses on the US and Euro servers?

As for the self-hosted Bitwarden Unified version I just tried, it has even reached 2024.2.0 but still does not support Inline Auto-Fill.

![2024-01-30_13-20-04](https://github.com/dani-garcia/vaultwarden/assets/119997727/879f8188-e9c2-45a8-b474-444b78b7a0d1)

**Reply by lukasj98** [reply]
Because I think it is a good feature

**Reply by bibicadotnet** [reply]
I think it will take a few more months for Bitwarden to perfect this feature
Currently, inline auto fill on Bitwarden is completely useless in my opinion

**Reply by BlackDex** [reply]
Again, it isn't the server which provides this feature, it's the clients.
The only thing the server tells the clients is that they can enable or disable the feature via those flags.
The clients read that endpoint.

So, the server can be any version, it will not work unless the clients support it or are triggered into using it (if they support that flag)
In the end, it's the client that needs to be v2024.2.0 probably, if they tend to bring that into that version.

**Reply by BlackDex** [reply]
Also, sometimes there are changes to be done on the server side for some of those feature flags, but that is only because those new features might use different server endpoints. Which is not the case for this feature, since that is totally a client side item.

**Reply by lukasj98** [reply]
And with which client version is it available

**Reply by BlackDex** [reply]
No idea. Thats up to Bitwarden.

**Reply by lukasj98** [reply]
But is it currently available?

**Reply by lukasj98** [reply]
?

**Reply by BlackDex** [reply]
It was available already 5 days ago.
https://github.com/dani-garcia/vaultwarden/releases/tag/1.30.2

**Reply by lukasj98** [reply]
> It was available already 5 days ago.
> 
> https://github.com/dani-garcia/vaultwarden/releases/tag/1.30.2
> 
> 

It is not available in the current release

**Reply by BlackDex** [reply]
It is, you just need to follow the directions as pointed out here already https://github.com/dani-garcia/vaultwarden/discussions/4254#discussioncomment-8171522

**Reply by stefan0xC** [reply]
Alternatively you can wait for the (recently released) `2024.2.0` version to hit your browser's add-on store before this feature will be available in the browser extension (i.e. without explicitly setting a feature flag). Cf. https://github.com/bitwarden/clients/commit/0127db44a00eaf27378d899d0c44aa255fd05a75

**Comment by ShlomiD83** [comment]
> Alternatively you can wait for the (recently released) `2024.2.0` version to hit your browser's add-on store before this feature will be available in the browser extension (i.e. without explicitly setting a feature flag). Cf. [bitwarden/clients@0127db4](https://github.com/bitwarden/clients/commit/0127db44a00eaf27378d899d0c44aa255fd05a75)

it's working great on 2024.2.0.
